From 0a9141c58a16974b614f4a5f392e7e789c95a2a1 Mon Sep 17 00:00:00 2001
From: Sergio Pena <sergio.pena@cloudera.com>
Date: Fri, 3 Jul 2015 10:51:34 -0700
Subject: [PATCH 149/176] =?UTF-8?q?PARQUET-152:=20Add=20validation=20on=20En?=
 =?UTF-8?q?coding.DELTA=5FBYTE=5FARRAY=20to=20allow=20FIX=E2=80=A6?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

PARQUET-152: Add validation on Encoding.DELTA_BYTE_ARRAY to allow FIXED_LEN_BYTE_ARRAY types.

  * FIXED_LEN_BYTE_ARRAY types are binary values that may use DELTA_BYTE_ARRAY encoding,
    so they should be allowed to be decoded using the same DELTA_BYTE_ARRAY encoding.

@rdblue @nezihyigitbasi  Could you review this fix?

I executed a test by writing a file that fall backs to DELTA_BYTE_ARRAY encoding, then read the file, and compare the read values with the written values, and it worked fine.

Author: Sergio Pena <sergio.pena@cloudera.com>

Closes #225 from spena/parquet-152 and squashes the following commits:

93fa03e [Sergio Pena] PARQUET-152: Add validation on Encoding.DELTA_BYTE_ARRAY to allow FIXED_LEN_BYTE_ARRAY types.

Conflicts:
	parquet-column/src/main/java/parquet/column/Encoding.java
Resolution:
    Fixed package names.
---
 .../src/main/java/parquet/column/Encoding.java     |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/parquet-column/src/main/java/parquet/column/Encoding.java b/parquet-column/src/main/java/parquet/column/Encoding.java
index 7aa5c1d..6a9dc4d 100644
--- a/parquet-column/src/main/java/parquet/column/Encoding.java
+++ b/parquet-column/src/main/java/parquet/column/Encoding.java
@@ -19,6 +19,7 @@
 package parquet.column;
 
 import static parquet.column.values.bitpacking.Packer.BIG_ENDIAN;
+import static parquet.schema.PrimitiveType.PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY;
 import static parquet.schema.PrimitiveType.PrimitiveTypeName.INT32;
 import static parquet.schema.PrimitiveType.PrimitiveTypeName.BINARY;
 import static parquet.schema.PrimitiveType.PrimitiveTypeName.BOOLEAN;
@@ -192,8 +193,8 @@ public enum Encoding {
     @Override
     public ValuesReader getValuesReader(ColumnDescriptor descriptor,
         ValuesType valuesType) {
-      if (descriptor.getType() != BINARY) {
-        throw new ParquetDecodingException("Encoding DELTA_BYTE_ARRAY is only supported for type BINARY");
+      if (descriptor.getType() != BINARY && descriptor.getType() != FIXED_LEN_BYTE_ARRAY) {
+        throw new ParquetDecodingException("Encoding DELTA_BYTE_ARRAY is only supported for type BINARY and FIXED_LEN_BYTE_ARRAY");
       }
       return new DeltaByteArrayReader();
     }
-- 
1.7.9.5

